### PR TITLE
Introducing Kiosk Mode, hiding certain menu items from the UI.

### PR DESCRIPTION
**Pull request by zigurana**
Hi!
This is the first part of introducing Kid / Kiosk mode into the RetroPie master branch for Emulationstation.
This PR introduces the hiding of menu items depending on the current UI mode. The second part will focus on hiding actual game-items. 
With this change, you can set the UI mode to "Full" (no changes), or "Kiosk" which will hide all menu items that can potentially alter the system configuration.
These include:
* In the start / GuiMenu: All items except for the "Quit" menu, which only contains the "shutdown system" option.
* In the select / GamelistOptions: the CustomCollection menus, and Edit Metadata

As was the case with the previous implementation, unlocking the protected mode (Kiosk), is done by inputting the konami cheat code (up, up, down, down, left, right, left, right, b, a).
This particular passphrase is stored in the settings as a plaintext string, if you would want to change it.

**Note**
>*There is NO filtering added in this PR, so no whitelisting/hiding of specific games, yet. Also, the RetroPie Menu is still available, altough there are other methods available to remove that if you want to lock down an installation.*

The next PR will hopefully bring those final pieces as well, altough this is a functional start in its own right.

---------------------------------------------
## Technical implementation
I've introduced a new nested class in Window: PassKeyListener, which stores the current progress, the passphrase, and the listenerfunction. I could split this out into a separate file, but it seemed overkill for such a small thing. Let me know if that would be appreciated.

Because I had to encapsulate a lot of the menu's inside the UI-mode test, I took the opportunity to refactor that single giant function into a collection of smaller methods, but there are no functional changes made there.

---------------------------------------------
## Testing
I've tested this as well as I could, but only on my windows build, so I would appreciate it if others would give this a spin on a Pi as well. 
I do not expect too many issues however, those I've kept for the next PR :-)

### CODE REVIEW

File: es-core/src/Window.cpp [hunk 1]
```diff
@@ -404,34 +409,83 @@ bool Window::isProcessing()
...
+		for(auto i = mGuiStack.begin(); i != mGuiStack.end(); i++)
+			(*i)->onScreenSaverDeactivate();
+	}
+}
+
+void Window::renderScreenSaver()
+{
+	if (mScreenSaver)
+		mScreenSaver->renderScreenSaver();
+}
```
Comment (joolswills) [inline es-core/src/Window.cpp]: Note it would be preferable to have these whitespace fixes above done as a separate commit in the PR (cleaner, more clear to see what is changed etc). These indent errors look like they were introduced with https://github.com/RetroPie/EmulationStation/commit/59d7516a1618604644067101aed87a4cc13f3fe0 - I must have missed them. tagging @pjft for reference.

**Comment by joolswills** [comment]
Thanks for this.

**Comment by PokeEngineer** [comment]
I like how you have implemented a Konami code easter egg into this. It fits well with the whole retro gaming theme of RetroPie!

**Comment by HerbFargus** [comment]
Looking good. Compiles successfully on my pi. I've tested it briefly and it hasn't crashed but I'll use it as my daily driver in the coming days just to see if anything comes up. I don't really use any of the fancy new doodads so I'm sure I probably won't cover some of the use cases.

A few minor notes/questions:

- You said filters weren't implemented yet, but they show up as an option in the menu, so I take it the menus are there but the implementation is not?

- Is it reasonable to have restart system as another option in the quit menu? Seems a bit redundant to have two levels of menus to select one option (though I'm not sure if there are going to be other menu options as you add other functions)

**Comment by zigurana** [comment]
@herb_fargus
The filtering you see now is already part of ES. 
Filtering as to be implemented by UI modes would not introduce new menu items, but rather enforce certain types silently. 

For Kiosk, that would mean filtering out all items with the hidden Metadata tag set to true (black-listing). 

For kids-mode, it would mean filtering out all items, except for the ones with the kid tag set to true (white-listing) .

Those filters would follow implicitly from setting the UI mode, and would not require any additional menu items. But again: it's all still to do.

Re. Quit item: I agree it is a bit ugly now. I will look into changing the item to a direct shutdown, rather than entering the submenu first. Do you think there is a use for 'restart'  in the kiosk mode?

**Comment by HerbFargus** [comment]
@zigurana  I guess that tells you how much I use emulationstation ;) good to know.

As far as the restart I'm fine either way, it's really an inconsequential cosmetic nit. Might be worth getting feedback from users on the forum that use ES more than me as to whether or not restart system should be part of the quit menu

**Comment by Darknior** [comment]
Thanks a lot, you make me happy with my children :D 
Will you in the future also add you "Kid Game" tag like now we have favorite ?

**Comment by zigurana** [comment]
@Darknior That is the plan yes, this is only the first step, but a useful addition on its own (I hope).

**Comment by Darknior** [comment]
Ok excellent, and thanks for the Konami code, so excellent idea to un lock the MENU without Keyboard ;)

File: es-core/src/Window.cpp (hunk 1 above)
Comment (pjft) [reply es-core/src/Window.cpp]: Thanks for tagging. I'll be mindful of those in the future - I believe this was before I had made sure to set up my IDE to handle the white spaces and tabs accordingly, but will double check, especially as I move to a new set up.

**Comment by pjft** [comment]
This is super, thanks for putting it together!

I am unable to test it for the moment, but looks good overall.

One thought: I agree it makes a lot of sense to restrict editing and all of that to the Full UI. In that sense, it might make sense to also edit BasicGameListView.cpp to remove the ability to edit favorites with the "Y" button, as well as remove the help text.

From reading the code, it seems that the "Start" menu only has an entry for the "Quit" menu, which then ends up only having "Shutdown".

While I have not significant arguments for or against having "Restart" as an option (though I wouldn't object to it in any way), you may end up having the "Start" button straight up call "openQuitMenu()" (well, the equivalent) when in KioskMode. Maybe that'd be an option to avoid going through the barebones menu somehow.

I'll likely make this my default mode, as I don't need the majority of the other menus in normal day to day usage.

Thanks!

**Comment by zigurana** [comment]
Some small improvements, mostly to facilitate the next step for this activity.
Also squashed the commits into one.

If there are no more comments, this is ready to be merged.

**Comment by zigurana** [comment]
@herb_fargus, did you get a chance to test this some more? 

@pjft, @jdrassa, @hex, @joolswils : if (some of) you could do a review of the proposed changes, I would be most obliged!

**Comment by pjft** [comment]
I would be thrilled to do so. 

I'll try to get a proper work setup this week - at least a laptop for ease of browsing and reading - and will try to look into this. 

Thanks for putting this together.

**Comment by zigurana** [comment]
Sorry for the (now deleted) premature message, I had not seen that there are merge conflicts after yesterday's commit. I'll fix that asap.

**Comment by zigurana** [comment]
done

File: es-app/src/guis/GuiMenu.h [hunk 2]
```diff
@@ -16,8 +16,16 @@ class GuiMenu : public GuiComponent
 
 private:
 	void addEntry(const char* name, unsigned int color, bool add_arrow, const std::function<void()>& func);
-	void openScreensaverOptions();
+	void addVersionInfo();
 	void openCollectionSystemSettings();
+	void openConfigInput();
```
Comment (pjft) [inline es-app/src/guis/GuiMenu.h]: Thank you for this very much needed split. GuiMenu was becoming quite unmanageable.

File: es-app/src/views/ViewController.cpp [hunk 3]
```diff
@@ -499,6 +499,28 @@ void ViewController::reloadAll()
...
+	{	
+		mCurUIMode = uimode;
+		reloadAll();
+		goToStart();
+	}
+}
+
+std::vector<std::string> ViewController::getUIModes()
+{
+	std::vector<std::string> uimodes = { "Full", "Kiosk" };
```
Comment (pjft) [inline es-app/src/views/ViewController.cpp]: I don't know enough about the future plans for this, but for clarity would this be better off as an enum, so that you don't have to do hardcoded text comparisons for what are well-defined terms up front? It's a minor thing, but worth asking. See also "isUIModeFull".

File: es-core/src/Window.cpp [hunk 4]
```diff
@@ -404,34 +409,75 @@ bool Window::isProcessing()
...
+
+void Window::PassKeyListener::listen(InputConfig * config, Input input, Window* window)
+{
+	// This function reads the current input to listen for the passkey
+	// sequence to unlock the UI mode. The progress is saved in mPassKeyCounter
+	// supported sequence-inputs: u (up), d (down), l (left), r (right), A, B, X, Y
```
Comment (pjft) [inline es-core/src/Window.cpp]: Small documentation nitpick - unless these are not case sensitive, I'd replace "A, B, X, Y" by their lower case counterparts :)

**Comment by pjft** [comment]
No concerns from a code perspective. Will try to compile this and give it a whirl today :) 

Thanks for putting this together!

File: es-app/src/guis/GuiMenu.cpp [hunk 5]
```diff
@@ -23,381 +23,412 @@
...
+		addEntry("UI SETTINGS", 0x777777FF, true, [this] {openUISettings(); });
+
+		addEntry("GAME COLLECTION SETTINGS", 0x777777FF, true, [this] { openCollectionSystemSettings(); });
+
+		addEntry("OTHER SETTINGS", 0x777777FF, true, [this] {openOtherSettings(); });
+
+		addEntry("CONFIGURE INPUT", 0x777777FF, true, [this] {openConfigInput(); });
+	}
+
+	addEntry("QUIT", 0x777777FF, true, [this] {openQuitMenu(); });
```
Comment (hex007) [inline es-app/src/guis/GuiMenu.cpp]: Whitespace inconsistency; space before curly bracket missing in some of these >  `[this] {openXXX(); }`

**Comment by hex007** [comment]
@zigurana I think UI settings and Sound settings should be available even in kiosk mode. If not UI atleast sound should be quickly accessible.

**Comment by pjft** [comment]
@hex007 I certainly don't oppose the Sound settings, though I think it all boils down to what @zigurana intends that mode to be.

If it's a mode that's supposed to be kind of "locked down" for others not to mess things up, I wouldn't add the audio there (or, if added, I would not save the setting if changed, always loading from the one in the config file). 

But, from that perspective, I also wonder if we shouldn't remove the ability to toggle favorites in it, as one can still mess things up that way.

If it's a mode for us to use on a day to day basis (which I would certainly like to have, even if it's not this but a mode with another name), I would support having sound settings available as well as the favorites (as it currently is).

File: es-app/src/views/ViewController.cpp (hunk 3 above)
Comment (zigurana) [reply es-app/src/views/ViewController.cpp]: But would that also mean we store the UIMode entry in settings as an enum (int)? I would rather have the settings remain user-readable, than prevent a string comparison.

Comment (pjft) [reply es-app/src/views/ViewController.cpp]: Ah. I suppose that's a good argument for keeping it as strings then.

Would we perhaps declare it in the .h for it to be easy to reference, or at least not inside a function if it's meant to have a meaning outside of it?

Either way is fine, just probing.

Thanks.

**Comment by zigurana** [comment]
Well, in my mind, the kiosk mode is a day-to-day usage mode. What I want to prevent is exposing options that can really mess up your installation (So UI settings, scraper etc).

So the real question is, what functions do you require to be **directly** accessible during your day-to-day usage? Remember that switching back to `Full` takes only about 5-30 sec (depending on the time needed to repopulate the GameListViews). For me, those are only:
* add/remove favs
* set the sound volume (I'll fix this asap).

Am I missing some more?

In the next step, Kids UI-Mode I plan to further lock down the system, including the volume and setting of favorites.

File: es-app/src/guis/GuiMenu.cpp (hunk 5 above)
Comment (zigurana) [reply es-app/src/guis/GuiMenu.cpp]: noted

**Comment by pjft** [comment]
Nope, that sounds about right then!

File: es-core/src/Window.cpp (hunk 4 above)
Comment (zigurana) [reply es-core/src/Window.cpp]: noted

**Comment by pjft** [comment]
A small detail: when you trigger the cheat code to go back to Full Mode, the event to open the gamelist view will still trigger, meaning that the video will start playing for a brief moment. At least with OMXPlayer that is the case, as it's a separate process.

Maybe you could catch that and prevent the animation and opening the gamelist from triggering. If not, it's not something that will happen _that_ often, but it's worth noting.

Otherwise it seems to work as intended. 

Thank you.

**Comment by zigurana** [comment]
Well spotted! Now you mention it, I think I've seen this as well, but never consciously picked this up. What I could do is:
~~1. move the `mPassKeyListener->listen()` method up so it's called before the `this->peekGui()->input(config, input);`
2. extend the mPassKeyListener to include some kind of flag-member that is set when a UI mode change is set.~~
1. Change the return type of `mPassKeyListener->listen()` to bool, indicating if a UI mode is changed, and
2. Only call `this->peekGui()->input(config, input);` when this is not the case.

that should work, I think?

**Comment by pjft** [comment]
I think that would work. You'd need to move it up, so it's called before the peekGui one.

File: es-app/src/guis/GuiGamelistOptions.cpp [hunk 6]
```diff
@@ -64,7 +65,9 @@ GuiGamelistOptions::GuiGamelistOptions(Window* window, SystemData* system) : Gui
 	mMenu.addRow(row);
 
 	std::map<std::string, CollectionSystemData> customCollections = CollectionSystemManager::get()->getCustomCollectionSystems();
-	if((customCollections.find(system->getName()) != customCollections.end() && CollectionSystemManager::get()->getEditingCollection() != system->getName()) ||
+
+	if(ViewController::get()->isUIModeFull() &&
```
Comment (pjft) [inline es-app/src/guis/GuiGamelistOptions.cpp]: So, here's an unintended behavior.

While I'd be happy to consider "add/remove" from a collection the same functionality as "favorites", I understand if you want to keep just "favorites" for the Kiosk mode. I won't object one way or the other.

Still, for what you currently intend to do, you need to add an extra set of brackets to encompass 

```
(customCollections.find(system->getName()) != customCollections.end() && CollectionSystemManager::get()->getEditingCollection() != system->getName()) || CollectionSystemManager::get()->getCustomCollectionsBundle()->getName() == system->getName()
``` 

otherwise it will work for the "My Collections" bundle but not for any other collection at the top level :) Just ran into this and thought I'd give you the heads up.

I'm using this as my main ES for the time being, so I'll let you know if anything else comes up :)

Comment (zigurana) [reply es-app/src/guis/GuiGamelistOptions.cpp]: Thanks for this! My intention was indeed to keep add/remove from collection **out** of the kiosk mode. I will add the brackets to fix it for all collection types.

**Comment by hex007** [comment]
The sound settings should be available so that people who don't know anything about kiosk mode can adjust it. if I leave it on kisosk mode I wouldn't want family members calling about adjusting volume. 

kiosk mode is used by more than those who know about it.

**Comment by HerbFargus** [comment]
>I wouldn't want family members calling about adjusting volume.

If I left it with a family member they would use a remote to turn the volume down. I've personally never used the volume in ES.

**Comment by hex007** [comment]
I have a portable handheld that has no external volume control. ES has one so I never felt the need. I did send over PRs to handle usb audio volume control just or that.

**Comment by pjft** [comment]
For cabinets and systems without sound control, the one in ES controls the entire system. On my living room setup I control it via the TV remote, but I expect this to be useful in a day to day basis for those setups.

**Comment by zigurana** [comment]
Yesyesyes, I said I would do it, didn't I? :-)

Enabled in kiosk mode:
* Volume control
* Shut down System
* Restart System
* Gamepad-shortcut for toggling favs

Am I missing something?

**Comment by zigurana** [comment]
>@zigurana I would also like to see kiosk mode being triggered temporarily and disabled via command line so people don't have to mess with key combo.

>Eg if triggered with a parameter the key combo should not work to get out of kiosk mode. This would be similar to --force_kiosk or something. This would prevent unauthorized access of settings

I could implement something similar to the ` --debug` flag?

**Comment by hex007** [comment]
That would be great. "--force-kiosk" sounds descriptive enough.

**Comment by zigurana** [comment]
Addressed all review comments, updated PR.

**Comment by hex007** [comment]
@zigurana 
- Tested the flag 
- Works as expected
- Exits on key combo
-  Combo is disabled as intended when forced through flag

**Comment by zigurana** [comment]
Are there any more concerns I need to address for this PR?

**Comment by hex007** [comment]
@zigurana I dont think so.

**Comment by pjft** [comment]
I haven't had the chance to review things since my last one but all the
things I mentioned at the time were addressed.

Thanks!

**Comment by pjft** [comment]
One small thing which I'm not sure is present in the last changes: when you switch to Kiosk mode, you would do well in finishing editing any current collection, otherwise Y will be locked to "add/remove games to <currently editing collection>". Unless that is a possible use case :)

**Comment by zigurana** [comment]
That is actually an edge case I had not yet considered, so thanks for that.
However, I think it's a low-occurence and low-impact side effect; the session locking into the edit mode for a custom collection rather than favorites.

Seeing that I have a number of other changes planned for the collection-management in the next delivery for UI modes, I'd rather leave it as is for now. 

Are you OK with that?

**Comment by pjft** [comment]
I have no concerns with it anyway. In fact, I could see some uses for it -
setting it to edit the kids collection and then lock it on kiosk mode for
them to just change their own collection rather than my favorites.

I was just flagging it as I ran into it today. :)

**Comment by joolswills** [comment]
Note that I will not be including this in 4.3 for RPI installs - final ES binaries are built, changelog is done, I just need to do a few bits and pieces and testing.

Also - unrelated to this PR but:
I might include a ES stable/dev module, and have it so we keep all development on another branch. Although issues have been fixed relatively quickly, there has been a few breakages, and it's an important part of retropie.

**Comment by zigurana** [comment]
Thanks for the heads-up.

A stable /dev split seems like a good idea. Do you need me to do another PR on the dev branch, or can you pull this PR into it?

**Comment by joolswills** [comment]
I will probably keep master as dev, and have a stable branch. But for now I would like to wait a few days after RetroPie 4.3 before this is included (for people that see the RetroPie 4.3 release and do source installs - eg on non rpi platforms).

**Comment by Joemommasfat** [comment]
I just tried out this branch.  Nice work!  One suggestion.  Can we disable the "settings" menu in kiosk mode?  By this I mean the retropie entry in the systems ribbon.  A lot of things can be broken in there if you start opening setup scripts.

**Comment by zigurana** [comment]
@Joemommasfat , thanks for testing this out!
Regarding hiding the RetroPie menu (system, actually) from the SystemView is part of the second stage of this activity. This stage is really just about closing off the actual EmulationStation menus.

BTW, the second stage is about finished, so once this gets accepted, I can just about open the next PR immediately!

**Comment by hex007** [comment]
@zigurana if this is ready to be merged, you might want to inform buzz about it.

**Comment by joolswills** [comment]
I know about it - see my post above. RetroPie 4.3 is just out, and I was hoping to do the emulationstation dev module before merging this.

**Comment by zigurana** [comment]
Hey, no worries / hurry, I have plenty of other stuff on my plate! Whenever the stable/dev split is in place is soon enough for me.

**Comment by joolswills** [comment]
I have switched the ES module to use a new "stable" branch now.

I have not yet finalised the new dev module, as I need to make some other changes (found some issues with the way we have emulationstation and emulationstation-kids in regards to installing/uninstalling).

Will sort that in the coming days, but I am happy to now merge this and anything else that is waiting. Thanks for your work on this.

**Comment by zigurana** [comment]
Thanks! Re: the old EmulationStation-kids branch, that is largely deprecated by now, or st least hopelessly outdated. I would not think it opportune to spend much time on fixing that mechanism (if kid is the only use case for that).

**Comment by pjft** [comment]
Thanks both - for developing this, and for merging it together with the stable branch work.

Just to confirm, then, if we now want to install and run this particular version of ES, we need to compile from source from the master branch, correct?

Is the plan to have binaries only for the stable branch?

Happy Friday everyone.

**Comment by joolswills** [comment]
@zigurana It also affected the emulationstation-dev module I was doing - but I just tested emulationstation-kids and I see the branch is gone, so it might be best just to remove that module now - if you are happy with that.

**Comment by zigurana** [comment]
Whoops, sorry for that (some overzealous cleaning from my part). Please remove the module,  the branch was very outdated anyways.
Thanks!

**Comment by joolswills** [comment]
@pjft there will only be binaries for the RPI for the stable branch - correct. You will need to use a source install using the upcoming emulationstation-dev module for the master branch.

**Comment by pjft** [comment]
Thanks, sounds good.

Out of curiosity, what is the plan for the "graduation" process, from master to stable? Do you already have anything in mind?

Let us know if you'd need any help with that, and thanks for taking the lead on this.

Best.

**Comment by joolswills** [comment]
No more than after testing - and the ability to easily revert without breaking the development flow.

**Comment by qphoria** [comment]
Can we get the option for "Restart Emulation Station" back in kiosk mode?

**Comment by qphoria** [comment]
Actually I'd like to recommend getting rid of Kid and Kiosk and just have Full and Controlled
Kid and Kiosk are only 2 preset entries of many possible settings. Would be best to have full control where you choose which menu items to show for your "controlled" situation. That would make both cleaner code and give more customization.

In my setup I'd like Kiosk users to be able to restart ES, set favorites, change the screensaver settings, change the theme. But disallow exit to prompt, edit metadata, running scraper, and various other settings. Would be nice to have them all customizable from the settings. Even if they were just manual entries in es_settings.xml
